### ISSUE TITLE
TF_CPP_MIN_LOG_LEVEL does not work with TF2.0 dev20190820

### ISSUE THREAD

**Issue by KichangKim** [tensorflow/tensorflow#31870]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 x64 1903
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.0.0-dev20190820
- Python version: 3.6.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA 10, cuDNN 7.6.2.24
- GPU model and memory: Geforce RTX 2080 ti 11GB

**Describe the current behavior**
Setting TF_CPP_MIN_LOG_LEVEL does not work lateset TF 2.0. If I set TF_CPP_MIN_LOG_LEVEL to 2, sill TF shows information and warning logs (including libpng warnings)

With dev20190504, this issue does not occurred.

**Describe the expected behavior**
Set TF_CPP_MIN_LOG_LEVEL=2 should prevent showing information/warning logs including libpng warnings.

**Code to reproduce the issue**
```
import os
import tensorflow as tf
import numpy as np

os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2"

a = tf.Variable(np.array([0, 1, 2]))

print(a)
```
You can see many [I] log using dev20190820. But with dev20190504, no logs.

**Comment by gadagashwini** [comment]
@KichangKim I tried on Colab with Tensorflow 2.0.0-dev20190820. Please see the colab [gist](https://colab.research.google.com/drive/12nzwoqnttwa4KFXR-xKs2MlnaDDWVghQ) and let us know the expected result. Thanks!

**Comment by KichangKim** [comment]
@gadagashwini Thanks for testing. But gist's result differ to my local result.
Here is my output:
```
2019-08-22 20:52:13.256379: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_100.dll
2019-08-22 20:52:19.121304: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library nvcuda.dll
2019-08-22 20:52:19.391093: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1618] Found device 0 with properties:
name: GeForce RTX 2080 Ti major: 7 minor: 5 memoryClockRate(GHz): 1.755
pciBusID: 0000:06:00.0
2019-08-22 20:52:19.398066: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_100.dll
2019-08-22 20:52:19.410017: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_100.dll
2019-08-22 20:52:19.419415: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cufft64_100.dll
2019-08-22 20:52:19.426158: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library curand64_100.dll
2019-08-22 20:52:19.450504: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusolver64_100.dll
2019-08-22 20:52:19.501850: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusparse64_100.dll
2019-08-22 20:52:19.545794: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2019-08-22 20:52:19.551803: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1746] Adding visible gpu devices: 0
2019-08-22 20:52:19.555811: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2019-08-22 20:52:19.564630: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1618] Found device 0 with properties:
name: GeForce RTX 2080 Ti major: 7 minor: 5 memoryClockRate(GHz): 1.755
pciBusID: 0000:06:00.0
2019-08-22 20:52:19.569926: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_100.dll
2019-08-22 20:52:19.573802: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_100.dll
2019-08-22 20:52:19.578127: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cufft64_100.dll
2019-08-22 20:52:19.582843: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library curand64_100.dll
2019-08-22 20:52:19.593490: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusolver64_100.dll
2019-08-22 20:52:19.598530: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusparse64_100.dll
2019-08-22 20:52:19.604157: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2019-08-22 20:52:19.610050: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1746] Adding visible gpu devices: 0
2019-08-22 20:52:21.725773: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1159] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-08-22 20:52:21.735777: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1165]      0
2019-08-22 20:52:21.738297: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1178] 0:   N
2019-08-22 20:52:21.742246: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1304] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 8686 MB memory) -> physical GPU (device: 0, name: GeForce RTX 2080 Ti, pci bus id: 0000:06:00.0, compute capability: 7.5)
<tf.Variable 'Variable:0' shape=(3,) dtype=int32, numpy=array([0, 1, 2])>
PS C:\Users\kicha\Downloads\tensorflow-issue>
```
Can you try my code on local machine?

**Comment by ziyigogogo** [comment]
same situation here, when i update my tf.nightly version to 2.0rc, TF_CPP_MIN_LOG_LEVEL does not work anymore.

**Comment by ziyigogogo** [comment]
After a quick dig into this, it seems that the problem appear only if you change the log level after import tensorflow. 
```python
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = "2"
import tensorflow as tf
import numpy as np
a = tf.Variable(np.array([0, 1, 2]))
print(a)
```
output:
```python
<tf.Variable 'Variable:0' shape=(3,) dtype=int32, numpy=array([0, 1, 2])>
```
if i change set the environ after import tf:
```python
import os
import tensorflow as tf
os.environ['TF_CPP_MIN_LOG_LEVEL'] = "2"
import numpy as np
a = tf.Variable(np.array([0, 1, 2]))
print(a)
```
output:
```python
...
2019-08-26 11:45:17.907378: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_100.dll
2019-08-26 11:45:20.776347: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library nvcuda.dll
2019-08-26 11:45:20.885617: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1618] Found device 0 with properties: 
name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.6575
pciBusID: 0000:01:00.0
2019-08-26 11:45:20.886163: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1618] Found device 1 with properties: 
name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.6575
pciBusID: 0000:02:00.0
2019-08-26 11:45:20.886391: I tensorflow/stream_executor/platform/default/dlopen_checker_stub.cc:25] GPU libraries are statically linked, skip dlopen check.
2019-08-26 11:45:20.887824: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1746] Adding visible gpu devices: 0, 1
2019-08-26 11:45:20.888138: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports 
...
```
This bug happens in Tensorflow 2.0rc, and in Tensorflow beta0&beta1 everything works good no matter the order of import.

I think the code somehow only the logger reading the TF_CPP_MIN_LOG_LEVEL when init in the latest version, and in the older version, the logger is reading TF_CPP_MIN_LOG_LEVEL on the fly. Could you please confirm if this is intend to do or a bug introduced in the latest version?
@gadagashwini

**Comment by ziyigogogo** [comment]
@ymodak 
![1](https://user-images.githubusercontent.com/9452949/63738157-7b9f7c80-c8bb-11e9-90a7-6948b5efbb1a.png)
tried 2.0.0-dev20190826, still not work. Maybe the platform-differ issues?
I'm testing this on win10.

**Comment by bersbersbers** [comment]
I am not sure this is related. I am using `2.0.0rc0`, but it seems the same issue is in `2.0.0.a0`. My code is 

```
import tensorflow as tf

tf.get_logger().setLevel('ERROR')
print("Hi!")
output = tf.keras.layers.Conv1D(1, 1)(tf.keras.layers.Input([1, 1]))
```
(I found this code in #28577, #28578).

and the output includes:
```
Hi!
2019-08-28 10:57:14.647006: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
```

If that is not the same issue, I can file it separately.

**Comment by ziyigogogo** [comment]
> I am not sure this is related. I am using `2.0.0rc0`, but it seems the same issue is in `2.0.0.a0`. My code is
> 
> ```
> import tensorflow as tf
> 
> tf.get_logger().setLevel('ERROR')
> print("Hi!")
> output = tf.keras.layers.Conv1D(1, 1)(tf.keras.layers.Input([1, 1]))
> ```
> 
> (I found this code in #28577, #28578).
> 
> and the output includes:
> 
> ```
> Hi!
> 2019-08-28 10:57:14.647006: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
> ```
> 
> If that is not the same issue, I can file it separately.
@bersbersbers 
I think its not the same issue, tf.get_logger().setLevel('ERROR') means set your python tensorflow log level to "ERROR", and the evn value "TF_CPP_MIN_LOG_LEVEL" is to set tensorflow c++ log level.
So in my code, i usually set them both like this:
"""python
 if args.debug is True:
        os.environ['TF_CPP_MIN_LOG_LEVEL'] = "0"
        tf.get_logger().setLevel("INFO")
else:
        os.environ['TF_CPP_MIN_LOG_LEVEL'] = "2"
        tf.get_logger().setLevel("ERROR")
"""

**Comment by ziyigogogo** [comment]
@ymodak any update?

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by ziyigogogo** [comment]
> It has been 15 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

yes it is.

**Comment by mihaimaruseac** [comment]
There is a new RC. Can you try with that one?

**Comment by ziyigogogo** [comment]
@mihaimaruseac still an issue. In case it is an IDE caused issue, i also tried with console
# this one still print the log
python -c "import os;import tensorflow as tf;os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2';import numpy as np;print(tf.Variable(np.array([0,1,2])))"
# this one will not print the log
python -c "import os;os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2';import tensorflow as tf;import numpy as np;print(tf.Variable(np.array([0,1,2])))"

**Comment by mihaimaruseac** [comment]
So I think you have to set the envvar before importing tf.

**Comment by ziyigogogo** [comment]
> So I think you have to set the envvar before importing tf.

Yes if the team think the change was on purpose.

For my previous user case, accept the param passed by argparse like this:


```python
import os
import tensorflow as tf
import argparse
def some_tf_fn():
    ...

if __main__ == "__name__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--log", type=bool, required=False, default=False)
    args = parser.parse_args()

    if args.log:
        os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
        tf.get_logger().setLevel("INFO")
    else:
        os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2"
        tf.get_logger().setLevel("ERROR")
```
In tf2.0-beta1, i can do things like this, but now due to the change i should re-orgnize my previous code. That's not a big deal, just want some confirmation from you guys since the change is not mentioned in any document. @mihaimaruseac

**Comment by mihaimaruseac** [comment]
I think it is some side effect on the interaction between absl and tensorflow.

**Comment by ziyigogogo** [comment]
OK, fair enough, i personally have no further question. thank you very much for the patient follow up though.

**Comment by mihaimaruseac** [comment]
I will try to investigate this further in my 20% time.

**Comment by bersbersbers** [comment]
> So I think you have to set the envvar before importing tf.

To add one more data points:
Following the official example at https://github.com/NVIDIA/tensorflow-determinism, the `TF_DETERMINISTIC_OPS` envvar can be set after importing `tensorflow`:

```
import tensorflow as tf
import os
os.environ['TF_DETERMINISTIC_OPS'] = '1'
# Now build your graph and train it
```

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by KichangKim** [comment]
@mihaimaruseac Thanks for investigation. So in my understand, this change is intended and will not be fixed, right? (if it is correct, I'll close this issue.)

**Comment by bersbersbers** [comment]
> So in my understand, this change is intended

I wonder what part of the discussion above makes you think that. @mihaimaruseac interpreted this as a side effect (not an intended effect) of some change, and has not concluded otherwise since mentioning to investigate this further.

**Comment by KichangKim** [comment]
@bersbersbers Oh, I misread ziyigogogo's comment. It seems that not yet concluded. Reopen.

**Comment by KichangKim** [comment]
Why the bot labeled this issue as "awaiting response"? It should be clearly labeled as "awaiting tensorflow".

**Comment by akhil189** [comment]
Hi. Is the issue solved?
If yes, could someone tell me how to avoid INFO and WARNING messages in output?

**Comment by KichangKim** [comment]
> Hi. Is the issue solved?
> If yes, could someone tell me how to avoid INFO and WARNING messages in output?

@akhil189 It seems not fixed yet. Temporary workaround is that call `os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2" ` before `import tensorflow as tf`, or manually set environment variable TF_CPP_MIN_LOG_LEVEL=2 before running your python script.
```
> SET TF_CPP_MIN_LOG_LEVEL=2
> python your_script.py
```

**Comment by akhil189** [comment]
@KichangKim Thank you for providing the workaround. But, I am using google colab and it's not working in that. The 'INFO' and 'WARNING' messages are still being shown in the output.

**Comment by awal-ahmed** [comment]
First import os and then set TF_CPP_MIN_LOG_LEVEL. After that import tensorflow. Like this:

import os
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2"

import tensorflow as tf

I think this will work fine.

**Comment by abal6725** [comment]
Hello, was there any resolution to this issue? I am having the same issue.

**Comment by anugrahasinha** [comment]
This issue seem to be still open, but with [workaround](https://github.com/tensorflow/tensorflow/issues/31870#issuecomment-553269261) mentioned by @KichangKim works fine.
Version : `2.3.0`

Sample for verification
```
$ python36
Python 3.6.2 (v3.6.2:5fd33b5, Jul  8 2017, 04:57:36) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
2020-09-08 04:26:59.370871: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'cudart64_101.dll'; dlerror: cudart64_101.dll not found
2020-09-08 04:26:59.379059: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
```

```
$ python36
Python 3.6.2 (v3.6.2:5fd33b5, Jul  8 2017, 04:57:36) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.environ['TF_CPP_MIN_LOG_LEVEL'] = '1'
>>> import tensorflow as tf
2020-09-08 04:27:33.738961: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'cudart64_101.dll'; dlerror: cudart64_101.dll not found
```

I know this is not an ideal env. I am using cygwin, on top of that I am calling a python binary installed for windows. Anyways, I think it does prove that the workaround for logging works.

**Comment by sushreebarsa** [comment]
I tried this in TF nightly version,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/8f62c3d79013d639850ffd5f4415a860/untitled101.ipynb?authuser=1)....thanks !

**Comment by sushreebarsa** [comment]
Closing this issue as it is fixed in latest version of TensorFlow. Please feel free to reopen the issue if you still have a concern. Thanks!

**Comment by SergioG-M** [comment]
I am getting warnings "WARNING:tensorflow:The following Variables were used in a Lambda layer's call" even if I set the env var. Using tf.get_logger().setLevel('ERROR') works though. Anyone else with the same issue?
